picard/script.py

```diff
@@ -910,6 +910,159 @@ def func_is_video(parser):
+    else:
+        return ''
+    multi_var = _get_multi_values(parser, multi, separator)
+    try:
+        return str(multi_var[index])
+    except IndexError:
+        return ''
+
+
+def func_foreach_multi(parser,
+                       multi,
+                       loop_code,
+                       separator=MULTI_VALUED_JOINER):
+    """Iterates over each element found in the specified multi-value variable.
+
+    Iterates over each element found in the specified multi-value variable, executing the specified code.
+    For each loop, the element value is first stored in the tag specified by _loop_value and the count is
+    stored in the tag specified by _loop_count.  This allows the element or count value to be accessed within
+    the code script.
+
+    Arguments:
+        multi: The multi-value to be iterated.
+        loop_code: Script code to be processed on each iteration.
+    """
+    loop_count = 0
+    multi_value = _get_multi_values(parser, multi, separator)
+    for index, value in enumerate(multi_value):
+        loop_count = index + 1
+        func_set(parser, '_loop_count', str(loop_count))
+        func_set(parser, '_loop_value', str(value))
+        loop_code.eval(parser)
+    func_unset(parser, '_loop_count')
+    func_unset(parser, '_loop_value')
+    return ''
+
+
+def func_while_loop(parser, condition, loop_code):
+    """Standard 'while' loop.  Also includes a runaway check to limit the maximum number of iterations.
+
+    Arguments:
+        condition: Script code to check before each iteration through the loop.
+        loop_code: Script code to be processed on each iteration.
+    """
+    if condition and loop_code:
+        runaway_check = 1000
+        loop_count = 0
+        while condition.eval(parser) and loop_count < runaway_check:
+            loop_count += 1
+            func_set(parser, '_loop_count', str(loop_count))
+            loop_code.eval(parser)
+        func_unset(parser, '_loop_count')
+    return ''
+
+
+def func_map_multi(parser,
+                   multi,
+                   loop_code,
+                   separator=MULTI_VALUED_JOINER):
+    """Iterates over each element found in the specified multi-value variable and updates the value.
+
+    Iterates over each element found in the specified multi-value variable and updates the value of the
+    element to the value returned by the specified code. For each loop, the element value is first stored in
+    the tag specified by _loop_value and the count is stored in the tag specified by _loop_count.  This
+    allows the element or count value to be accessed within the code script.
+
+    Arguments:
+        multi: The multi-value to be iterated.
+        loop_code: Script code to be processed on each iteration.
+
+    Returns the updated multi-value variable.
+    """
+    loop_count = 0
+    multi_value = _get_multi_values(parser, multi, separator)
+    for index, value in enumerate(multi_value):
+        loop_count = index + 1
```

Comment:
Same remark: `enumerate(multi_value, 1)` and `multi_value[loop_count - 1] = ...` will simplify code a bit.

Reply:
Will change.